Aliased packages are incorrectly resolved

### Verify latest release

- [X] I verified that the issue exists in the latest pnpm release

### pnpm version

9.0.2

### Which area(s) of pnpm are affected? (leave empty if unsure)

Dependencies resolver

### Link to the code that reproduces this issue or a replay of the bug

_No response_

### Reproduction steps

- Run `pnpm add ui-kit@npm:@org/repo@x.x.x`, where `@org/repo@x.x.x` is any private package (the only important thing is to use version which doesn't match any of existing versions of the public `ui-kit` package).
- Observe an error complaining that there is no such version of the `ui-kit` package

### Describe the Bug

We have a private package aliased as `ui-kit` the following way:

```
"ui-kit": "npm:@org/repo@x.x.x",
```
After migrating to pnpm 9, installing a new version of any package results in the following error:
```
ERR_PNPM_NO_MATCHING_VERSION  No matching version found for ui-kit@x.x.x`
This error happened while installing a direct dependency of ...

The latest release of ui-kit is "0.20.10".
```

If I get it correctly, there's a public `ui-kit` package, which has nothing to do with our private one.
However `pnpm` prefers this this package over the private one that we specified in `package.json` using alias.


### Expected Behavior

`pnpm` should respect aliases and correctly resolve private package without trying to resolve the public one first.

### Which Node.js version are you using?

20.12.1

### Which operating systems have you used?

- [X] macOS
- [ ] Windows
- [ ] Linux

### If your OS is a Linux based, which one it is? (Include the version if relevant)

_No response_

I've been experiencing this with aliasing `eslint-plugin-node` as `eslint-plugin-n`. For some reason, I cannot reproduce this locally on my machine, but I have been getting these errors when Renovate tries to do a PR.

[package.json](https://github.com/weareinreach/InReach/blob/dev/packages/eslint-config/package.json) entry: `"eslint-plugin-node": "npm:eslint-plugin-n@16.6.2",`

[Relevant PR comment](https://github.com/weareinreach/InReach/pull/1231#issuecomment-2063858919)

<details>
<summary>Error details from PR comment</summary>

```
Scope: all 15 workspace projects
Progress: resolved 1, reused 0, downloaded 0, added 0
Progress: resolved 56, reused 0, downloaded 0, added 0
Progress: resolved 123, reused 0, downloaded 0, added 0
Progress: resolved 154, reused 0, downloaded 0, added 0
Progress: resolved 186, reused 0, downloaded 0, added 0
/tmp/renovate/repos/github/weareinreach/InReach/packages/eslint-config:
 ERR_PNPM_NO_MATCHING_VERSION  No matching version found for eslint-plugin-node@16.6.2

This error happened while installing a direct dependency of /tmp/renovate/repos/github/weareinreach/InReach/packages/eslint-config

The latest release of eslint-plugin-node is "11.1.0".

Other releases are:
  * old: 4.2.3
  * next: 7.0.0-beta.0

If you need the full list of all 60 published versions run "$ pnpm view eslint-plugin-node versions".
```

</details>

<details>
<summary>Error message from Renovate job</summary>

```
DEBUG: rawExec err (branch="renovate/all-minor-patch")
{
  "baseBranch": "dev"
  "err": {
    "cmd": "/bin/sh -c pnpm install --recursive --lockfile-only --ignore-scripts --ignore-pnpmfile",
    "stderr": "",
    "stdout": "Scope: all 15 workspace projects\nProgress: resolved 1, reused 0, downloaded 0, added 0\nProgress: resolved 56, reused 0, downloaded 0, added 0\nProgress: resolved 123, reused 0, downloaded 0, added 0\nProgress: resolved 154, reused 0, downloaded 0, added 0\nProgress: resolved 186, reused 0, downloaded 0, added 0\n/tmp/renovate/repos/github/weareinreach/InReach/packages/eslint-config:\n ERR_PNPM_NO_MATCHING_VERSION  No matching version found for eslint-plugin-node@16.6.2\n\nThis error happened while installing a direct dependency of /tmp/renovate/repos/github/weareinreach/InReach/packages/eslint-config\n\nThe latest release of eslint-plugin-node is \"11.1.0\".\n\nOther releases are:\n  * old: 4.2.3\n  * next: 7.0.0-beta.0\n\nIf you need the full list of all 60 published versions run \"$ pnpm view eslint-plugin-node versions\".\n",
    "options": {
      "cwd": "/tmp/renovate/repos/github/weareinreach/InReach",
      "encoding": "utf-8",
      "env": {
        "NPM_CONFIG_CACHE": "/tmp/renovate/cache/others/npm",
        "npm_config_store": "/tmp/renovate/cache/others/pnpm",
        "HOME": "/home/ubuntu",
        "PATH": "/home/ubuntu/.cargo/bin:/home/ubuntu/.local/bin:/go/bin:/home/ubuntu/bin:/home/ubuntu/.npm-global/bin:/home/ubuntu/.cargo/bin:/home/ubuntu/.local/bin:/go/bin:/home/ubuntu/bin:/home/ubuntu/.npm-global/bin:/home/ubuntu/.cargo/bin:/home/ubuntu/.local/bin:/go/bin:/home/ubuntu/bin:/home/ubuntu/.npm-global/bin:/home/ubuntu/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin",
        "LC_ALL": "C.UTF-8",
        "LANG": "C.UTF-8",
        "CONTAINERBASE_CACHE_DIR": "/tmp/renovate/cache/containerbase"
      },
      "maxBuffer": 10485760,
      "timeout": 900000
    },
    "exitCode": 1,
    "name": "ExecError",
    "message": "Command failed: pnpm install --recursive --lockfile-only --ignore-scripts --ignore-pnpmfile\n",
    "stack": "ExecError: Command failed: pnpm install --recursive --lockfile-only --ignore-scripts --ignore-pnpmfile\n\n    at ChildProcess.<anonymous> (/usr/local/renovate/lib/util/exec/common.ts:99:11)\n    at ChildProcess.emit (node:events:529:35)\n    at ChildProcess.emit (node:domain:489:12)\n    at Process.ChildProcess._handle.onexit (node:internal/child_process:292:12)"
  }
  "durationMs": 5520
}

DEBUG: lock file error (branch="renovate/all-minor-patch")
{
  "baseBranch": "dev"
  "commands": [
    "pnpm install --recursive --lockfile-only --ignore-scripts --ignore-pnpmfile",
    "pnpm dedupe --config.ignore-scripts=true"
  ]
  "err": {
    "cmd": "/bin/sh -c pnpm install --recursive --lockfile-only --ignore-scripts --ignore-pnpmfile",
    "stderr": "",
    "stdout": "Scope: all 15 workspace projects\nProgress: resolved 1, reused 0, downloaded 0, added 0\nProgress: resolved 56, reused 0, downloaded 0, added 0\nProgress: resolved 123, reused 0, downloaded 0, added 0\nProgress: resolved 154, reused 0, downloaded 0, added 0\nProgress: resolved 186, reused 0, downloaded 0, added 0\n/tmp/renovate/repos/github/weareinreach/InReach/packages/eslint-config:\n ERR_PNPM_NO_MATCHING_VERSION  No matching version found for eslint-plugin-node@16.6.2\n\nThis error happened while installing a direct dependency of /tmp/renovate/repos/github/weareinreach/InReach/packages/eslint-config\n\nThe latest release of eslint-plugin-node is \"11.1.0\".\n\nOther releases are:\n  * old: 4.2.3\n  * next: 7.0.0-beta.0\n\nIf you need the full list of all 60 published versions run \"$ pnpm view eslint-plugin-node versions\".\n",
    "options": {
      "cwd": "/tmp/renovate/repos/github/weareinreach/InReach",
      "encoding": "utf-8",
      "env": {
        "NPM_CONFIG_CACHE": "/tmp/renovate/cache/others/npm",
        "npm_config_store": "/tmp/renovate/cache/others/pnpm",
        "HOME": "/home/ubuntu",
        "PATH": "/home/ubuntu/.cargo/bin:/home/ubuntu/.local/bin:/go/bin:/home/ubuntu/bin:/home/ubuntu/.npm-global/bin:/home/ubuntu/.cargo/bin:/home/ubuntu/.local/bin:/go/bin:/home/ubuntu/bin:/home/ubuntu/.npm-global/bin:/home/ubuntu/.cargo/bin:/home/ubuntu/.local/bin:/go/bin:/home/ubuntu/bin:/home/ubuntu/.npm-global/bin:/home/ubuntu/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin",
        "LC_ALL": "C.UTF-8",
        "LANG": "C.UTF-8",
        "CONTAINERBASE_CACHE_DIR": "/tmp/renovate/cache/containerbase"
      },
      "maxBuffer": 10485760,
      "timeout": 900000
    },
    "exitCode": 1,
    "name": "ExecError",
    "message": "Command failed: pnpm install --recursive --lockfile-only --ignore-scripts --ignore-pnpmfile\n",
    "stack": "ExecError: Command failed: pnpm install --recursive --lockfile-only --ignore-scripts --ignore-pnpmfile\n\n    at ChildProcess.<anonymous> (/usr/local/renovate/lib/util/exec/common.ts:99:11)\n    at ChildProcess.emit (node:events:529:35)\n    at ChildProcess.emit (node:domain:489:12)\n    at Process.ChildProcess._handle.onexit (node:internal/child_process:292:12)"
  }
  "type": "pnpm"
}
```

</details>

--- a/.changeset/two-points-peel.md
+++ b/.changeset/two-points-peel.md
@@ -1,0 +1,6 @@
+---
+"@pnpm/resolve-dependencies": patch
+"pnpm": patch
+---
+
+Fix aliased dependencies resolution on repeat install with existing lockfile [#7957](https://github.com/pnpm/pnpm/issues/7957).

--- a/pkg-manager/resolve-dependencies/src/resolveDependencies.ts
+++ b/pkg-manager/resolve-dependencies/src/resolveDependencies.ts
@@ -1139,7 +1139,11 @@ async function resolveDependency (
   }
   try {
     if (!options.update && currentPkg.version && currentPkg.pkgId?.endsWith(`@${currentPkg.version}`)) {
-      wantedDependency.pref = currentPkg.version
+      if (semver.valid(wantedDependency.pref)) {
+        wantedDependency.pref = currentPkg.version
+      } else if (wantedDependency.pref.startsWith('npm:')) {
+        wantedDependency.pref = `${wantedDependency.pref.substring(0, wantedDependency.pref.lastIndexOf('@') + 1)}${currentPkg.version}`
+      }
     }
     pkgResponse = await ctx.storeController.requestPackage(wantedDependency, {
       alwaysTryWorkspacePackages: ctx.linkWorkspacePackagesDepth >= options.currentDepth,

--- a/pkg-manager/resolve-dependencies/src/resolveDependencies.ts
+++ b/pkg-manager/resolve-dependencies/src/resolveDependencies.ts
@@ -1139,7 +1139,7 @@ async function resolveDependency (
   }
   try {
     if (!options.update && currentPkg.version && currentPkg.pkgId?.endsWith(`@${currentPkg.version}`)) {
-      if (semver.valid(wantedDependency.pref)) {
+      if (semver.validRange(wantedDependency.pref)) {
         wantedDependency.pref = currentPkg.version
       } else if (wantedDependency.pref.startsWith('npm:')) {
         wantedDependency.pref = `${wantedDependency.pref.substring(0, wantedDependency.pref.lastIndexOf('@') + 1)}${currentPkg.version}`

